### ISSUE TITLE
linux/mman.h include?

### ISSUE THREAD

**Issue by petrovr** [openssl/openssl#3900]
Secure arena (crypto/mem_sec.c) includes linux/mman.h.
Issue is that header linux/mman.h does not exist on Android for API < 21.
It seems to me on library builds fine without warnings.

What about to remove "#  include <linux/mman.h>" from crypto/mem_sec.c?

**Comment by kaduk** [comment]
That include is necessary to detect the definition of MLOCK_ONFAULT (if supported by the kernel in question); removing the include would remove support for MLOCK_ONFAULT.

**Comment by petrovr** [comment]
kaduk wrote:
> That include is necessary to detect the definition of MLOCK_ONFAULT (if supported by the kernel in question); removing the include would remove support for MLOCK_ONFAULT.

Ok.
On Android MLOCK_ONFAULT  is not defined in platform headers.  In 
"unified headers" is defined for some platform.
The I guess it could be excluded only for Android.

**Comment by richsalz** [comment]
How old is Android 21?

We do not want to clutter our code with ifdef for old platforms.

**Comment by petrovr** [comment]
Rich Salz wrote:
> How old is Android 21?
> We do not want to clutter our code with ifdef for old platforms.

For instance API level 19-20 correspond to Android KitKat 4.4 - 4.4.4 
and 4.4.4 is still supported (security patches).

**Comment by bernd-edlinger** [comment]
Is there a no-secure-memory configure option?

**Comment by levitte** [comment]
> Is there a no-secure-memory configure option?

Nope.  But that could certainly be arranged.

Another way to handle it could be to guard with a check for `__ANDROID__` and whatever macro there might be that indicated the API level.  Perhaps with an extra Android-specific include?

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Another way to handle it could be to guard with a check for `__ANDROID__` and whatever macro there might be that indicated the API level.  Perhaps with an extra Android-specific include?

Most of functionality is obtained from <sys/mman.h>. It seems to me on 
Android <linux/mman.h> could be excluded unconditionally.

**Comment by bernd-edlinger** [comment]
maybe <sys/mman.h> works in general ?
Also the linux man page says:
```
       #include <sys/mman.h>

       void *mmap(void *addr, size_t length, int prot, int flags,
                  int fd, off_t offset);

```

**Comment by petrovr** [comment]
Bernd Edlinger wrote:
> maybe <sys/mman.h> works in general ?
yes and no : Please see previous post - linux/mman.h is to found 
definition of MLOCK_ONFAULT.

**Comment by bernd-edlinger** [comment]
MLOCK_ONFAULT is currently 0x1 do you expect linux to ever change that ?

**Comment by levitte** [comment]
Well, that's not too hard to solve, is it?

```
: ; cc foo.c
: levitte@lapdog:~
: ; ./a.out 
1
: levitte@lapdog:~
: ; cat foo.c
#include <stdio.h>
#include <sys/mman.h>
#ifdef linux
# include <linux/mman.h>
#endif

int main()
{
  printf("%d\n", MLOCK_ONFAULT);
}
```

(perhaps replace `#ifdef linux` with something that is better tuned for Android's different APIs)

**Comment by petrovr** [comment]
Bernd Edlinger wrote:
> MLOCK_ONFAULT is currently 0x1 do you expect linux to ever change that ?
Wrong direction. Value itself is not important. Please read again post 
from "kaduk".

**Comment by bernd-edlinger** [comment]
If you use gcc the following can be used to find candidates:
`gcc -x c -g3 -E /dev/null |grep define`

so one of these could be used to detect "linux":
```
#define __gnu_linux__ 1
#define __linux 1
#define __linux__ 1
#define linux 1
#define __unix 1
#define __unix__ 1
#define unix 1
```

**Comment by richsalz** [comment]
Don’t go there.  ☺

**Comment by bernd-edlinger** [comment]
Maybe ifdef __ANDROID__ will work ?

**Comment by richsalz** [comment]
I mean don't try to add our own define for the MLOCK magic number :)
We can always add -DOPENSSL_SYS_ANDROID to our config definitions.

**Comment by bernd-edlinger** [comment]
but is andoid garanteed to never have any linux/mman.h ??

**Comment by richsalz** [comment]
(But I know @dot-asm prefers to use system-defined values if possible)

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Well, that's not too hard to solve, is it?

It is solved in my local repository as:
#if !defined(__ANDROID__)
#  include <linux/mman.h>
#endif

*linux* and *unix* are compiler defines declared for android targets as 
well.

**Comment by bernd-edlinger** [comment]
is this gcc?

**Comment by richsalz** [comment]
Does it matter?  Android is a phone O/S and MLOCK_ONFAULT seems not necessary there.

**Comment by bernd-edlinger** [comment]
@levitte  we already use that define:

```
crypto/dso/dso_dlfcn.c:        defined(__ANDROID__)
crypto/async/arch/async_posix.h:    && !defined(__ANDROID__) && !defined(__OpenBSD__)
include/openssl/rand.h:# if defined(__ANDROID__) && defined(__NDK_FPABI__)
```

**Comment by levitte** [comment]
Cool, so something like this? 

``` C
# if defined(linux) && !defined(__ANDROID__) 
```

**Comment by petrovr** [comment]
Richard Levitte wrote:
> Cool, so something like this?
>
> ``` C
> # if defined(linux) && !defined(__ANDROID__)
> ```
>
inclusion  is already in # if defined(OPENSSL_SYS_LINUX)
may be  defined(linux) is duplicate ?

**Comment by bernd-edlinger** [comment]
Yes, `linux` is duplicate, and `__ANDROID__` is probably enough.
Generally I think `__linux__` or `__linux` are more portable than `linux`
because without the underscore it is just an ordinary name that could
have no special meaning on another O/S.

**Comment by KangLin** [comment]
MLOCK_ONFAULT is defined in #include <asm-generic/mman.h>, then 
#include <asm/mman.h> include asm-generic/mman.h. 
The <asm/mman.h> is supported by the all linux kernel .
Why don't directly include <asm/mman.h>?

**Comment by petrovr** [comment]
Perhaps linux/mman.h -> asm/mman.h is best solution. 

For protocol issue could be resolved with use of android unified headers (available since ndk14+) but this opens another issue - minimum supported NDK version. // Out of #3900 scope.

**Comment by dot-asm** [comment]
Resolved in #5584.
